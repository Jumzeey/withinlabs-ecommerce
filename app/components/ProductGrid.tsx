'use client';

import { Product } from '@/app/types';
import { ProductCard } from './ProductCard';
import { motion } from 'framer-motion';

export function ProductGrid({ products }: { products: Product[] }) {
  return (
    <motion.div
      data-testid="product-grid"
      initial={{ opacity: 0 }}
      animate={{ opacity: 1 }}
      transition={{ duration: 0.5 }}
      className="grid grid-cols-1 gap-6 sm:grid-cols-2 lg:grid-cols-3 xl:grid-cols-4"
    >
      {products.map((product, index) => (
        <ProductCard
          key={product.id}
          product={product}
        />
      ))}
    </motion.div>
  );
}