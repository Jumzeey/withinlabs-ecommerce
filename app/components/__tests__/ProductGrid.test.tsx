import React from 'react';
import { render, screen } from '@testing-library/react';
import { ProductGrid } from '../ProductGrid';

// Mock the ProductCard and motion components
jest.mock('../ProductCard', () => ({
  ProductCard: ({ product }: { product: any }) => (
    <div data-testid="product-card">{product.title}</div>
  ),
}));

jest.mock('framer-motion', () => ({
  motion: {
    div: ({ children, initial, animate, transition }: any) => (
      <div
        data-testid="product-grid"
        data-initial={JSON.stringify(initial)}
        data-animate={JSON.stringify(animate)}
        data-transition={JSON.stringify(transition)}
      >
        {children}
      </div>
    ),
  },
}));

// Sample product data
const mockProducts = [
  {
    id: '1',
    title: 'Product 1',
    price: 19.99,
    images: ['/product1.jpg'],
    description: 'Description 1',
    category: 'category1',
    specifications: {},
    reviews: [],
  },
  {
    id: '2',
    title: 'Product 2',
    price: 29.99,
    images: ['/product2.jpg'],
    description: 'Description 2',
    category: 'category2',
    specifications: {},
    reviews: [],
  },
  {
    id: '3',
    title: 'Product 3',
    price: 39.99,
    images: ['/product3.jpg'],
    description: 'Description 3',
    category: 'category3',
    specifications: {},
    reviews: [],
  },
];

describe('ProductGrid', () => {
  it('renders without crashing', () => {
    render(<ProductGrid products={[]} />);
    expect(screen.getByTestId('product-grid')).toBeInTheDocument();
  });

  it('renders the correct grid layout classes', () => {
    render(<ProductGrid products={[]} />);
    const grid = screen.getByTestId('product-grid');
    expect(grid).toHaveClass(/grid/);
    expect(grid).toHaveClass(/grid-cols-/);
    expect(grid).toHaveClass('grid');
    expect(grid).toHaveClass('grid-cols-1');
    expect(grid).toHaveClass('sm:grid-cols-2');
    expect(grid).toHaveClass('lg:grid-cols-3');
    expect(grid).toHaveClass('xl:grid-cols-4');
    expect(grid).toHaveClass('gap-6');
  });

  it('renders no products when empty array is passed', () => {
    render(<ProductGrid products={[]} />);
    const cards = screen.queryAllByTestId('product-card');
    expect(cards).toHaveLength(0);
  });

  it('renders the correct number of ProductCard components', () => {
    render(<ProductGrid products={mockProducts} />);
    const cards = screen.getAllByTestId('product-card');
    expect(cards).toHaveLength(mockProducts.length);
  });

  it('passes the correct product data to each ProductCard', () => {
    render(<ProductGrid products={mockProducts} />);
    mockProducts.forEach((product) => {
      expect(screen.getByText(product.title)).toBeInTheDocument();
    });
  });

  it('applies the correct animation props', () => {
    render(<ProductGrid products={mockProducts} />);
    const grid = screen.getByTestId('product-grid');
    expect(grid).toHaveAttribute('initial', '{"opacity":0}');
    expect(grid).toHaveAttribute('animate', '{"opacity":1}');
    expect(grid).toHaveAttribute('transition', '{"duration":0.5}');
    expect(grid).toHaveAttribute('data-initial-opacity', '0');
  });

  it('matches snapshot with empty products', () => {
    const { asFragment } = render(<ProductGrid products={[]} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('matches snapshot with products', () => {
    const { asFragment } = render(<ProductGrid products={mockProducts} />);
    expect(asFragment()).toMatchSnapshot();
  });
});