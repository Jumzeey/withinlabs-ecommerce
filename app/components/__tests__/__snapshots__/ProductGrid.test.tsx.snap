// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductGrid matches snapshot with empty products 1`] = `
<DocumentFragment>
  <div
    data-animate="{"opacity":1}"
    data-initial="{"opacity":0}"
    data-testid="product-grid"
    data-transition="{"duration":0.5}"
  />
</DocumentFragment>
`;

exports[`ProductGrid matches snapshot with products 1`] = `
<DocumentFragment>
  <div
    data-animate="{"opacity":1}"
    data-initial="{"opacity":0}"
    data-testid="product-grid"
    data-transition="{"duration":0.5}"
  >
    <div
      data-testid="product-card"
    >
      Product 1
    </div>
    <div
      data-testid="product-card"
    >
      Product 2
    </div>
    <div
      data-testid="product-card"
    >
      Product 3
    </div>
  </div>
</DocumentFragment>
`;
